src/adapter/auth-oso-rebac/src/oso_dataset_authorizer.rs

```diff
@@ -0,0 +1,284 @@
+        {
+            Ok(allowed) if allowed => Ok(()),
+            Ok(_not_allowed) => Err(DatasetActionUnauthorizedError::Access(
+                AccessError::Forbidden(
+                    DatasetActionNotEnoughPermissionsError {
+                        action,
+                        dataset_ref: dataset_handle.as_local_ref(),
+                    }
+                    .into(),
+                ),
+            )),
+            Err(e) => Err(DatasetActionUnauthorizedError::Internal(e.int_err())),
+        }
+    }
+
+    #[tracing::instrument(level = "debug", skip_all, fields(%dataset_handle))]
+    async fn get_allowed_actions(
+        &self,
+        dataset_handle: &odf::DatasetHandle,
+    ) -> Result<HashSet<DatasetAction>, InternalError> {
+        let (user_actor, dataset_resource) =
+            try_join!(self.user_actor(), self.dataset_resource(dataset_handle))?;
+
+        self.kamu_auth_oso
+            .get_allowed_actions(user_actor, dataset_resource)
+            .int_err()
+    }
+
+    #[tracing::instrument(level = "debug", skip_all, fields(dataset_handles=?dataset_handles, action=%action))]
+    async fn filter_datasets_allowing(
+        &self,
+        dataset_handles: Vec<odf::DatasetHandle>,
+        action: DatasetAction,
+    ) -> Result<Vec<odf::DatasetHandle>, InternalError> {
+        let user_actor = self.user_actor().await?;
+        let mut matched_dataset_handles = Vec::with_capacity(dataset_handles.len());
+
+        let dataset_ids = dataset_handles
+            .iter()
+            .map(|hdl| hdl.id.clone())
+            .collect::<Vec<_>>();
+        let dataset_resources_resolution = self
+            .oso_resource_service
+            .get_multiple_dataset_resources(&dataset_ids)
+            .await
+            .int_err()?;
+        let mut dataset_handle_id_mapping =
+            dataset_handles
+                .into_iter()
+                .fold(HashMap::new(), |mut acc, hdl| {
+                    acc.insert(hdl.id.clone(), hdl);
+                    acc
+                });
+
+        for (dataset_id, dataset_resource) in dataset_resources_resolution.resolved_resources {
+            let is_allowed = self
+                .kamu_auth_oso
+                .is_allowed(user_actor.clone(), action, dataset_resource)
+                .int_err()?;
+
+            if is_allowed {
+                let dataset_handle = dataset_handle_id_mapping
+                    // Thus we obtain the value without cloning
+                    .remove(&dataset_id)
+                    .ok_or_else(|| {
+                        format!("Unexpectedly, dataset_handle was found: {dataset_id}").int_err()
+                    })?;
+
+                matched_dataset_handles.push(dataset_handle);
+            }
+        }
+
+        Ok(matched_dataset_handles)
+    }
+
+    #[tracing::instrument(level = "debug", skip_all, fields(dataset_handles=?dataset_handles, action=%action))]
+    async fn classify_datasets_by_allowance(
```

Comment:
Minor: maybe `classify_dataset_handles_by_allowance`, since we have a 2nd routine by ids now.